### PR TITLE
Extract sampling and learning logic into a separate method

### PR DESCRIPTION
**Pull request by mrsamsami**

### CODE REVIEW

File: hive/agents/dqn.py [hunk 1]
```diff
@@ -297,51 +297,52 @@ def update(self, update_info, agent_traj_state=None):
...
                 next_state_inputs,
                 batch,
             ) = self.preprocess_update_batch(batch)
 
-            # Compute predicted Q values
-            self._optimizer.zero_grad()
-            pred_qvals = self._qnet(*current_state_inputs)
-            actions = batch["action"].long()
-            pred_qvals = pred_qvals[torch.arange(pred_qvals.size(0)), actions]
+            # Update the agent based on it
```
Comment (hnekoeiq) [inline hive/agents/dqn.py]: Maybe a bit more elaborate comment?

File: hive/agents/drqn.py [hunk 2]
```diff
@@ -240,56 +240,57 @@ def update(self, update_info, agent_traj_state=None):
...
-            target_hidden_state = self._target_qnet.init_hidden(
-                batch_size=self._batch_size,
-            )
-            # Compute predicted Q values
-            self._optimizer.zero_grad()
-            pred_qvals, _ = self._qnet(*current_state_inputs, hidden_state)
-            pred_qvals = pred_qvals.view(self._batch_size, self._max_seq_len, -1)
-            actions = batch["action"].long()
-            pred_qvals = torch.gather(pred_qvals, -1, actions.unsqueeze(-1)).squeeze(-1)
+            # Update the agent based on it
```
Comment (hnekoeiq) [inline hive/agents/drqn.py]: same comment as above.

File: hive/agents/dqn.py [hunk 3]
```diff
@@ -297,51 +297,52 @@ def update(self, update_info, agent_traj_state=None):
...
+        # Update target network
+        if self._target_net_update_schedule.update():
+            self._update_target()
 
-            q_targets = batch["reward"] + self._discount_rate * next_qvals * (
-                1 - batch["done"]
-            )
+        return agent_traj_state
+
+    def _update_on_batch(self, batch, current_state_inputs, next_state_inputs):
```
Comment (hnekoeiq) [inline hive/agents/dqn.py:322]: Maybe a better name? Like `_update_qnet()`?

**Review by hnekoeiq — APPROVED** [review]
Some minor comments. Otherwise, LGTM.

File: hive/agents/dqn.py (hunk 3 above)
Comment (mrsamsami) [reply hive/agents/dqn.py:322]: Given that `_update_on_batch()` is shared by all agents, my suggestion would be to keep it as it is. Of course, if you have any strong objections to this reasoning, we can change it.

**Pull request closed without merging**